### PR TITLE
VP integration - create a DB table for the placements

### PR DESCRIPTION
**Pull request by ivukotic**
This addresses issue #4933.

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,3 @@
+{
```
Comment (bari12) [inline .vscode/settings.json]: Remove this file

File: lib/rucio/core/did_meta_plugins/did_column_meta.py [hunk 2]
```diff
@@ -127,7 +127,7 @@ def get_metadata(self, scope, name, session=None):
         except NoResultFound:
             raise exception.DataIdentifierNotFound("Data identifier '%(scope)s:%(name)s' not found" % locals())
 
-    def set_metadata(self, scope, name, key, value, recursive=False, session=None):
+    def set_metadata(self, scope, name, key, value, recursive=False, session=None):  # ILIJA
```
Comment (bari12) [inline lib/rucio/core/did_meta_plugins/did_column_meta.py]: Remove comment

File: lib/rucio/core/did_meta_plugins/json_meta.py [hunk 3]
```diff
@@ -46,7 +46,7 @@
 from rucio.db.sqla.util import json_implemented
 
 
-class JSONDidMeta(DidMetaPlugin):
+class JSONDidMeta(DidMetaPlugin):  # ILIJA
```
Comment (bari12) [inline lib/rucio/core/did_meta_plugins/json_meta.py]: Remove comment

File: testDIST.py [hunk 4]
```diff
@@ -0,0 +1,27 @@
+from math import asin, atan2, cos, radians, sin, sqrt
```
Comment (bari12) [inline testDIST.py]: Remove file

File: testGEOIP.py [hunk 5]
```diff
@@ -0,0 +1,10 @@
+import socket
```
Comment (bari12) [inline testGEOIP.py]: Rmove File

File: testHUGE.py [hunk 6]
```diff
@@ -0,0 +1,44 @@
+import xmltodict
```
Comment (bari12) [inline testHUGE.py]: Remove File

File: testSORTING.py [hunk 7]
```diff
@@ -0,0 +1,44 @@
+# UC
```
Comment (bari12) [inline testSORTING.py]: Remove File

File: lib/rucio/db/sqla/migrate_repo/versions/9a45bc4ea66d_add_vp_table.py [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+
+
+def upgrade():
+    '''
+    Upgrade the database to this revision
+    '''
+
+    if context.get_context().dialect.name in ['oracle', 'mysql', 'postgresql']:
+        create_table('did_virtual_placements',
+                     sa.Column('scope', sa.String(25)),
```
Comment (bari12) [inline lib/rucio/db/sqla/migrate_repo/versions/9a45bc4ea66d_add_vp_table.py]: sa.String(get_schema_value('SCOPE_LENGTH')))

File: lib/rucio/db/sqla/migrate_repo/versions/9a45bc4ea66d_add_vp_table.py [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+                     sa.Column('name', sa.String(255)),
```
Comment (bari12) [inline lib/rucio/db/sqla/migrate_repo/versions/9a45bc4ea66d_add_vp_table.py]: sa.String(get_schema_value('NAME_LENGTH')))

File: lib/rucio/db/sqla/migrate_repo/versions/9a45bc4ea66d_add_vp_table.py [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+down_revision = '739064d31565'
+
+
+def upgrade():
+    '''
+    Upgrade the database to this revision
+    '''
+
+    if context.get_context().dialect.name in ['oracle', 'mysql', 'postgresql']:
+        create_table('did_virtual_placements',
```
Comment (bari12) [inline lib/rucio/db/sqla/migrate_repo/versions/9a45bc4ea66d_add_vp_table.py]: Let's just call it `virtual_placements` scratch the did.

File: lib/rucio/db/sqla/migrate_repo/versions/9a45bc4ea66d_add_vp_table.py [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+                     sa.Column('scope', sa.String(25)),
+                     sa.Column('name', sa.String(255)),
+                     sa.Column('placements', JSON()),
+                     sa.Column('created_at', sa.DateTime),
+                     sa.Column('updated_at', sa.DateTime)
+                     )
+
+        create_primary_key('DID_VP_PK', 'did_virtual_placements', ['scope', 'name'])
```
Comment (bari12) [inline lib/rucio/db/sqla/migrate_repo/versions/9a45bc4ea66d_add_vp_table.py]: --> VP_PK

File: lib/rucio/db/sqla/migrate_repo/versions/9a45bc4ea66d_add_vp_table.py [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+        create_foreign_key('DID_VP_FK', 'did_virtual_placements', 'dids',
```
Comment (bari12) [inline lib/rucio/db/sqla/migrate_repo/versions/9a45bc4ea66d_add_vp_table.py]: VP_DIDS_FK

File: lib/rucio/db/sqla/models.py [hunk 13]
```diff
@@ -461,6 +465,17 @@ class DataIdentifier(BASE, ModelBase):
                    Index('DIDS_EXPIRED_AT_IDX', 'expired_at'))
 
 
+class DidVirtualPlacements(BASE, ModelBase):
+    """Represents virtual placements"""
+    __tablename__ = 'did_virtual_placements'
+    scope = Column(InternalScopeString(get_schema_value('SCOPE_LENGTH')))
```
Comment (bari12) [inline lib/rucio/db/sqla/models.py:471]: See changes in migrate script

**Review by bari12 — CHANGES_REQUESTED** [review]
Also remove the `GeoLite2-City.mmdb` file please

**Comment by ivukotic** [comment]
that was a massive mess up on my side. I should be more careful with the -a option. Now all of the comments should be addressed.

**Comment by bari12** [comment]
Hi Ilija,

This looks GTG, can you just rebase it please. I think that's also why the tests fail.
Also please squash the commits into one (I can do it this time, just for the next one)
Please also update https://github.com/rucio/rucio/blob/master/etc/sql/oracle/schema.sql (In a separate PR) and get this information into there too.

**Comment by ivukotic** [comment]
Hi Martin,

I merged upstream changes but it still fails... now with logger thing?

Ilija

**Comment by bari12** [comment]
Ah this was an issue on master just yesterday, it is fixed, please rebase again. Sorry

**Comment by ivukotic** [comment]
seems now it works. can you please merge it?

**Review by bari12 — APPROVED** [review]
All good, please add it in a separate PR to https://github.com/rucio/rucio/blob/master/etc/sql/oracle/schema.sql

**Pull request merged**